luatest/justrun.lua

```diff
@@ -0,0 +1,164 @@
+--   return value
+--
+-- - quote_args (boolean, default: false)
+--
+--   Quote CLI arguments before concatenating them into a shell
+--   command.
+--
+-- @string dir Directory where the process will run.
+-- @table env Environment variables for the process.
+-- @table args Options that will be passed when the process starts.
+-- @table opts Custom options: nojson, stderr and qoute_args.
+-- @return table
+function justrun.tarantool(dir, env, args, opts)
+    assert(type(dir) == 'string')
+    assert(type(env) == 'table')
+    assert(type(args) == 'table')
+    opts = opts or {}
+    assert(type(opts) == 'table')
+
+    -- Prevent system/user inputrc configuration file from
+    -- influencing testing code.
+    env['INPUTRC'] = '/dev/null'
+
+    local tarantool_exe = arg[-1]
+    -- Use popen.shell() instead of popen.new() due to lack of
+    -- cwd option in popen (gh-5633).
+    local env_str = table.concat(fun.iter(env):map(function(k, v)
+        return ('%s=%q'):format(k, v)
+    end):totable(), ' ')
+    local args_str = table.concat(fun.iter(args):map(function(v)
+        return opts.quote_args and ('%q'):format(v) or v
+    end):totable(), ' ')
+    local command = ('cd %s && %s %s %s'):format(dir, env_str, tarantool_exe,
+                                                 args_str)
+    log.info('Running a command: %s', command)
+    local mode = opts.stderr and 'rR' or 'r'
+    local ph = popen.shell(command, mode)
+
+    local stderr_fiber
+    if opts.stderr then
+        stderr_fiber = collect_stderr(ph)
+    end
+
+    -- Read everything until EOF.
+    local chunks = {}
+    while true do
+        local chunk, err = ph:read()
+        if chunk == nil then
+            cancel_stderr_fiber(stderr_fiber)
+            ph:close()
+            error(err)
+        end
+        if chunk == '' then -- EOF
+            break
+        end
+        table.insert(chunks, chunk)
+    end
+
+    local exit_code = ph:wait().exit_code
+    local stderr = join_stderr_fiber(stderr_fiber)
+    ph:close()
+
+    -- If an error occurs, discard the output and return only the
+    -- exit code. However, return stderr.
+    if exit_code ~= 0 then
+        return {
+            exit_code = exit_code,
+            stderr = stderr,
+        }
+    end
+
+    -- Glue all chunks, strip trailing newline.
+    local res = table.concat(chunks):rstrip()
+    log.info('Command output:\n%s', res)
+
+    -- Decode JSON object per line into array of tables (if
+    -- `nojson` option is not passed).
+    local decoded
+    if opts.nojson then
+        decoded = res
+    else
+        decoded = fun.iter(res:split('\n')):map(json.decode):totable()
+    end
+
+    return {
+        exit_code = exit_code,
+        stdout = decoded,
+        stderr = stderr,
+    }
+end
+
+-- `popen` module requires Tarantool 2.4.1 and newer
+-- Otherwise justrun module will be an empty table.
+return has_popen and justrun or {}
```

Comment:
This should be changed if you consider my comments above.

Reply:
Yep, fixed. (comment also deleted)